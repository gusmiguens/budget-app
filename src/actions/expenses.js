import { v4 as uuidv4 } from 'uuid'
import database from '../firebase/firebase'

export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
})

export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
})

export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
})

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '',
            note = '',
            amount = '',
            createdAt = 0
        } = expenseData
        const expense = {
            description,
            note,
            amount,
            createdAt
        }
        return database.ref('expenses').push(expense)
            .then((snapshot) => {
                dispatch(addExpense({
                    id: snapshot.key,
                    ...expense
                }))
            })
    }
}

/*
How it worked:
    - component calls action generator
    - action generator returns object
    - component dispatches object
    - redux store changes
*/

/*
How it works with firebase
    - component calls action generator
    - action generator returns function
    - component dispatches function (?)
    - function runs (has the ability to dispatch other actions and do whatever it wants
*/