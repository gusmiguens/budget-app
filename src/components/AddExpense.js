const React = require('react')
const { connect } = require('react-redux')
const ExpenseForm = require('./ExpenseForm')
const { addExpense } = require('../actions/expenses')

class AddExpense extends React.Component {
    onSubmit = (expense) => {
        this.props.addExpense(expense)
        this.props.history.push('/')
    }
    render() {
        return (
            <div>
                <h1>Add Expense</h1>
                <ExpenseForm
                    onSubmit={this.onSubmit}
                />
            </div>
        )
    }
}

const mapDispatchToProps = (dispatch) => ({
    addExpense: (expense) => dispatch(addExpense(expense))
})

const connectedAddExpense = connect(undefined, mapDispatchToProps)(AddExpense)

module.exports = {
    connectedAddExpense,
    AddExpense
}