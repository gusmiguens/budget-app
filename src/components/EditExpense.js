const React = require('react')
const { connect } = require('react-redux')
const { editExpense, removeExpense } = require('../actions/expenses')
const ExpenseForm = require('./ExpenseForm')

class EditExpense extends React.Component {
    onSubmit = (expense) => {
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push('/')
    }
    onRemove = () => {
        this.props.removeExpense({ id: this.props.expense.id });
        this.props.history.push('/')
    }
    render() {
        return (
            <div>
                <ExpenseForm
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />
                <button 
                onClick={this.onRemove}
                >Remove Expense
                </button>
            </div>
        )
    }
}

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
})

const mapDispatchToProps = (dispatch, props) => ({
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    removeExpense: (data) => dispatch(removeExpense(data))
})

const connectedEditExpense = connect(mapStateToProps, mapDispatchToProps)(EditExpense)

module.exports = {
    connectedEditExpense,
    EditExpense
}