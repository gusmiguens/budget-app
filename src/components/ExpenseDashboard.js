const React = require('react')
const { ExpenseList } = require('./ExpenseList')
const { ExpenseListFilters } = require('./ExpenseListFilters')

const ExpenseDashboard = () => (
    <div>
        <ExpenseListFilters />
        <ExpenseList />
    </div>
)

module.exports = ExpenseDashboard