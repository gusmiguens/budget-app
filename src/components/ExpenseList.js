const React = require('react')
const { connect } = require('react-redux')
const ExpenseListItem = require('./ExpenseListItem')
const selectExpenses = require('../selectors/expenses')

const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (
                <p>No expenses</p>
            ) : (
                    props.expenses.map((expense) => {
                        return <ExpenseListItem key={expense.id} {...expense} />
                    })
                )
        }
    </div>
)

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
}

const ConnectedExpenseList = connect(mapStateToProps)(ExpenseList)

module.exports = {
    ConnectedExpenseList,
    ExpenseList
}