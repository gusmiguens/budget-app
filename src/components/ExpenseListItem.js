const React = require('react')
const { Link } = require('react-router-dom')

const ExpenseListItem = ({ description, amount, createdAt, id }) => (
    <div>
        <Link to={`/edit/${id}`}>
            <h3>{description}</h3>
        </Link>
        <p>Amount: {amount} - CreatedAt: {createdAt}</p>
    </div>
)

module.exports = ExpenseListItem