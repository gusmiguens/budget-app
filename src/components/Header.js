import React from 'react'
import { NavLink } from 'react-router-dom'

export default () => (
    <header>
        <h1>Budget</h1>
        <NavLink to='/' activeClassName='is-active' exact={true}>Dashboard</NavLink>
        <NavLink to='/create' activeClassName='is-active'>Create Page</NavLink>
    </header>
)