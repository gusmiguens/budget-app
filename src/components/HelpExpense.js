import React from 'react'

export default () => (
    <div>Help Expense</div>
)