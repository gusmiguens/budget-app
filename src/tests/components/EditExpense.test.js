import React from 'react'
import { shallow } from 'enzyme'
import { EditExpense } from '../../components/EditExpense'
import expenses from '../fixtures/expenses'

let editExpense, removeExpense, history, wrapper
beforeEach(() => {
    editExpense = jest.fn()
    removeExpense = jest.fn()
    history = { push: jest.fn() }
    wrapper = shallow(
        <EditExpense
            editExpense={editExpense}
            removeExpense={removeExpense}
            history={history}
            expense={expenses[1]}
        />
    )
})

describe('<EditExpense />', () => {
    it('Should render EditExpense', () => {
        expect(wrapper).toMatchSnapshot()
    })
    it('Should handle editExpense', () => {
        wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1])
        expect(history.push).toHaveBeenLastCalledWith('/')
        expect(editExpense).toHaveBeenLastCalledWith(expenses[1].id, expenses[1])
    })
    it('Should render removeExpense', () => {
        wrapper.find('button').simulate('click')
        expect(history.push).toHaveBeenLastCalledWith('/')
        expect(removeExpense).toHaveBeenLastCalledWith({
            id: expenses[1].id
        })
    })
})