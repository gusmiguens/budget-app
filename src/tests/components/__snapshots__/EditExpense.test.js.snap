// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditExpense /> Should render EditExpense 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 2000,
          "createdAt": -345600000,
          "description": "Rent",
          "id": "2",
          "note": "",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="button button--secondary"
      onClick={[Function]}
    >
      Remove expense
    </button>
  </div>
</div>
`;
